fondant/pipeline.py

```diff
@@ -0,0 +1,311 @@
+"""This module defines classes to represent a Fondant Pipeline."""
+import logging
+import json
+import yaml
+import typing as t
+from dataclasses import dataclass
+from pathlib import Path
+
+from fondant.component import FondantComponentSpec
+from fondant.component_spec import ComponentSubset
+from fondant.import_utils import is_kfp_available
+from fondant.exceptions import InvalidPipelineDefinition
+
+if is_kfp_available():
+    import kfp.dsl as dsl
+    import kfp
+    from kubernetes import client as k8s_client
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class FondantComponentOperation(FondantComponentSpec):
+    """
+    Class representing an operation for a Fondant Component in a Kubeflow Pipeline. An operation
+    is a representation of a function that will be executed as part of a pipeline.
+
+    Arguments:
+        component_spec_path: The path to the specification file defining the component
+        arguments: A dictionary containing the argument name and value for the operation.
+        number_of_gpus: The number of gpus to assign to the operation
+        node_pool_name: The name of the node pool to which the operation will be assigned.
+        p_volumes: Collection of persistent volumes in a Kubernetes cluster. Keys are mount paths,
+         values are Kubernetes volumes or inherited types(e.g. PipelineVolumes).
+        ephemeral_storage_request = Used ephemeral-storage request (minimum) for the operation.
+         Defined by string which can be a number or a number followed by one of “E”, “P”, “T”, “G”,
+         “M”, “K”. (e.g. 2T for 2 Terabytes)
+
+    Note:
+        - A Fondant Component operation is created by defining a Fondant Component and its input
+            arguments.
+        - The `number_of_gpus`, `node_pool_name`,`p_volumes` and `ephemeral_storage_size`
+            attributes are optional and can be used to specify additional configurations for
+            the operation. More information on the optional attributes that can be assigned to
+             kfp components here:
+             https://kubeflow-pipelines.readthedocs.io/en/1.8.13/source/kfp.dsl.html
+    """
+
+    component_spec_path: str
+    arguments: t.Dict[str, t.Any]
+    number_of_gpus: t.Optional[int] = None
+    node_pool_name: t.Optional[str] = None
+    p_volumes: t.Optional[t.Dict[str, k8s_client.V1Volume]] = None
+    ephemeral_storage_size: t.Optional[str] = None
+
+    def __post_init__(self):
+        with open(self.component_spec_path, encoding="utf-8") as file_:
+            specification = yaml.safe_load(file_)
+        super().__init__(specification)
+
+
+class FondantPipeline:
```

Comment:
This class seems to be a combination of a `Client` and a `Pipeline`. I would either split it, or abstract the `Client` away even more.

Eg. the following methods don't really make sense to me, but could be converted:
- `pipeline.get_pipeline_id(pipeline_name)`  --> `pipeline.id`
- `pipeline.get_pipeline_version_ids(pipeline_id)` --> `pipeline.version_ids`
- `pipeline.compile_pipeline(..., components, ...)` --> `pipeline.compile(...)

It would be good to start from the user interface of the pipeline and work backwards.

Reply:
Thanks for the feedback. I went with the approach of separating both classes since it's the one that makes the most sense. The class and examples are now updated